crates/sparrow-runtime/src/prepare/execute_iter.rs

```diff
@@ -0,0 +1,538 @@
+        };
+
+        // Find which indices to take from the batch (the non-late data)
+        let take_indices = time_column
+            .iter()
+            .enumerate()
+            .filter_map(|(index, time)| {
+                let time = time.expect("valid time");
+                if time >= self.watermark {
+                    self.watermark = std::cmp::max(self.watermark, time - self.bounded_lateness);
+                    Some(index as u64)
+                } else {
+                    // Late data - drop it
+                    None
+                }
+            })
+            .collect::<Vec<u64>>();
+        let take_indices: PrimitiveArray<UInt64Type> =
+            PrimitiveArray::from_iter_values(take_indices);
+
+        // Drop all rows with late data from each column
+        let filtered_columns = unfiltered_batch
+            .columns()
+            .iter()
+            .map(|column| arrow::compute::take(column, &take_indices, None))
+            .try_collect()
+            .into_report()
+            .change_context(Error::PreparingColumn)?;
+        let record_batch = RecordBatch::try_new(unfiltered_batch.schema(), filtered_columns)
+            .into_report()
+            .change_context(Error::PreparingColumn)?;
+
+        // 2. Slicing may reduce the number of entities to operate and sort on.
+        let record_batch = self.slice_preparer.slice_batch(record_batch)?;
+
+        // 3. Prepare each of the columns by getting the column behavior result
+        // let prepared_columns: Vec<ArrayRef> = self
+        //     .columns
+        //     .iter_mut()
+        //     .map(|column| column.get_result(None, Some(&self.key_hash_inverse), &record_batch))
+        //     .try_collect()?;
+
+        let mut prepared_columns: Vec<ArrayRef> = Vec::new();
+        for c in self.columns.iter_mut() {
+            let batch = c
+                .get_result(None, Some(&self.key_hash_inverse), &record_batch)
+                .await?;
+            prepared_columns.push(batch);
+        }
+
+        let record_batch = RecordBatch::try_new(self.prepared_schema.clone(), prepared_columns)
+            .into_report()
+            .change_context(Error::PreparingColumn)?;
+
+        // 4. After preparing the batch, concatenate the leftovers from the previous batch
+        // Note this is done after slicing, since the leftovers were already sliced.
+        let record_batch = if let Some(leftovers) = self.leftovers.take() {
+            debug_assert!(self.leftovers.is_none());
+            arrow::compute::concat_batches(&self.prepared_schema, &[leftovers, record_batch])
+                .into_report()
+                .change_context(Error::PreparingColumn)?
+        } else {
+            // The only time there should be no leftovers is the first batch, but we still need to handle this case.
+            record_batch
+        };
+
+        // 5. Pull out the time, subsort, and key hash columns to sort the record batch
```

Comment:
I wonder if this could be implemented as a wrapper around the existing prepare iterator, rather than as an alternative. What I'm picturing:

1. Run prepare to produce batches
2. Use a custom iterator, or something like `fold` to maintain state, and produce batches using the logic here.

The benefit is that (1) we wouldn't duplicate the prepare logic -- only one place to fix bugs (2) it would be easier to understand what *this* does different.

Reply:
The `execute_iter` contains the input buffer logic. Currently, part of that logic is "dropping late data", where late data is defined as "data that arrives with an event time prior to the current watermark". The watermark is defined as the current largest event time minus an allowed lateness delta. 

The problem if we use `prepare` then fold over batches produced by that iterator is we lose _ordering_, and we cannot tell if we should've dropped data. For example, if the following messages are read by the stream:
```
(allowed lateness=5)
[0, 2, 3, 5, 7, 10] - no late data, don't drop any
[0, 10, 2, 3, 5, 7] - drop [2, 3]. 
```
But since `prepare` sorts the batch of messages, the logic around the iterator would only see the ordered batch, and not know it should've dropped `[2, 3`. This is a problem because we then get non-deterministic results based on batch size.

Reply:
I think even once we're iterated on how we handle late data and have the ability to walk back entities, we'll still want a cutoff time that drops late data. For example, if we're reading events in 2023, and we see an event from 2022, it likely is no longer relevant for making a prediction right now, and we shouldn't re-process a year's worth of events for such an outlier.

Reply:
That's a good catch. What it seems like is we actually have three things happening -- we probably don't need to implement it now, but:

1. The underlying *reader*. Reading batches from Parquet, or reading rows from Pulsar and assembling batches. This is probably where the late data handling *should* happen. Maybe watermark too.
2. The general *preparation*. This could be a map or fold over the batches from the reader. In this case, it would do things like prepare the columns, sort things batches if not done by the reader, etc.
3. Some special stuff. Not sure we'd still need execute/prepare to be separate.

But, I think you're argument is convincing for why handling late data / watermarks needs to happen *before* the more general things.

Reply:
Added a comment for this